submissions/xT0PERx/Nanodegree-arcade-game/app.js

```diff
@@ -0,0 +1,102 @@
+const X_START_POSITION_PLAYER = 200;
+const Y_START_POSITION_PLAYER = 395;
+const X_START_POSITION_ENEMY = 0;
+const Y_START_POSITION_ENEMY = [55, 140, 225];
+const SPEED_ENEMY = [100, 200, 300];
+const STEP_LEFT_RIGHT = 100;
+const STEP_UP_DOWN = 85;
+const MAX_STEP_PLAYER = 400;
+const MIN_LENGTH_FIELD = 0;
+const MAX_LENGTH_FIELD = 500;
+const HALF_LENGTH_ENEMY = 50;
+
+let countWin = 0;
+// Enemies our player must avoid
+var Enemy = function (x, y, speed, objPlayer) {
+  this.x = x;
+  this.y = y;
+  this.speed = speed;
+  this.objPlayer = objPlayer;
+  this.sprite = "images/enemy-bug.png";
+};
+// Update the enemy's position, required method for game
+// Parameter: dt, a time delta between ticks
+Enemy.prototype.update = function (dt) {
+  this.x += this.speed * dt;
+  if (this.x > MAX_LENGTH_FIELD) {
+    this.x = -STEP_LEFT_RIGHT;
+    this.speed = Math.random() * MAX_LENGTH_FIELD + HALF_LENGTH_ENEMY;
+  }
+
+  this.collide();
+};
+// Draw the enemy on the screen, required method for game
+Enemy.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Enemy.prototype.collide = function () {
+  if (
+    this.objPlayer.y === this.y &&
+    this.objPlayer.x < this.x + HALF_LENGTH_ENEMY &&
+    this.objPlayer.x > this.x - HALF_LENGTH_ENEMY
+  ) {
+    this.objPlayer.goStartPosition();
+  }
+};
+const Player = function (x, y) {
+  this.x = x;
+  this.y = y;
+  this.sprite = "images/char-boy.png";
+};
+Player.prototype.update = function (dt) {};
+
+Player.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Player.prototype.handleInput = function (arr) {
+  if (arr === "left" && this.x > MIN_LENGTH_FIELD) {
+    this.x -= STEP_LEFT_RIGHT;
+  } else if (arr === "right" && this.x < MAX_STEP_PLAYER) {
+    this.x += STEP_LEFT_RIGHT;
+  } else if (arr === "up") {
+    this.y -= STEP_UP_DOWN;
+    if (this.y < MIN_LENGTH_FIELD) {
+      player.goStartPosition();
+      alert("Wins: " + ++countWin);
+    }
+  } else if (arr === "down" && this.y < Y_START_POSITION_PLAYER) {
+    this.y += STEP_UP_DOWN;
+  }
+};
+
+Player.prototype.goStartPosition = function () {
+  this.x = X_START_POSITION_PLAYER;
+  this.y = Y_START_POSITION_PLAYER;
+};
+
+const player = new Player(X_START_POSITION_PLAYER, Y_START_POSITION_PLAYER);
+
+const countEnemies = 3;
+const allEnemies = Array(countEnemies)
+  .fill()
+  .map(function (enemy, index) {
```

Comment:
Your IDE/linter should have warned you that `enemy` is an unused variable.
Call back signature promises we may expect that the array by that moment contains elements, each of which represents an enemy. In fact it is not. The array contains elements that are `undefined` each. It is misleading. Imagine this `.map` doesn't follow `.fill()` immediately but instead an array of undefineds is passed over to some function defined in a completely different part of code base and gets mapped there. A peer developer will spend some time trying to debug and understand why all elements are `undefined` while they are expected to be something that identifies an enemy.

In a nutshell, this `.map` here doesn't transform incoming array elements into something new but rather replaces them with entities that do not depend on incoming array elements at all.

`Array#map` only makes sense where outcoming array elements depend on incoming array elements.

The only useful thing here is index or rather an iteration through a sequence of integers starting with `0` limited with a certain count.

This is a clear use case to employ the `for` operator. `Array#push` is still a legal method to use in cases like this.

Reply:
There is yet another approach to avoid `Array#push` similar to what you had initially, yet not really elegant as well but more obvious:
`const someEntities = Array.from({length: numberOfEntitiesToCreate}, (_, i) => new Entity(parameters, i, evenMoreParameters))`
In fact it is equivalent to `Array().fill().map()` yet reads better since it is all-in-one.
Although this approach is questionnable.

Reply:
Thanks a lot, good full explanation of "why it is better not to do this". I will try to continue to write code first of all for people and for ease of maintenance of the code in the future, and not just for it to work.